### PR TITLE
Revolver Tweaks... more like additions, Gunslinger quirk, and more

### PR DESCRIPTION
**Pull request by rye-rice**
## About The Pull Request

Oh god. What started as a simple project to fix a bug where ammo boxes acted like speedloaders and revolver safties turned into. A lot.
In short:
- Single action revolvers now have the hammer as a saftey. Techinically they still have none, but decocking the hammer works as one.
- Double actions get a proper safety. Yayy
- Cleans up some behavior related to ammo boxes. I need to fucking clean up that whole file someday.
- Revolvers... should feel better to use. They no longer fast load from everything now, and are slightly faster to load than magazines
- You probably saw the number of changes in revolver.dm. Yes, along with being less janky, revolvers support gate loading revolvers. This is mostly for the shadow and ashhand. Other revolvers still drop everything on unload.
- You have new actions to choose from when using revovlers. Alt click to access.
- Adds support to remove energy gun cells like magazines. Leftover from the branch i poached gun huds from
- OH YEAH, GUN HUDS! Stolen from skyrat https://github.com/Skyrat-SS13/Skyrat-tg/pull/4333 but made much less shittier and with sprites made by azlan. This technically isn't used yet, maybe in the future. It was mostly modified to be used with revolvers, to see which chambers are loaded
- The gunslinger quirk! See below. 0 points.
- This "resprites" the shadow, but i'd just call it making it better proportioned compared to others. Also adds several more skins.
- This fixes a fuckup with the flaming arrow as well.

To elaborate on gunslinger:

It gives you more recoil and spread when using SMGs, assault rifles, and LMGs, it increases accuracy with revolvers, shotguns, and lever actions, and no changes to anything else.

Also, lets you fan SINGLE ACTION revolvers (HP Shadow, HP Montague, and HP Ashhand) by one handing it with nothing else in your offhand. *Slightly* faster than pistols to boot.

Sound effect for moving the cyllinder is https://freesound.org/people/shelbyshark/sounds/501560/ by the way. Not from TGMC this time, suprisingly

## Why It's Good For The Game

Should make friendly fire incidents involving revolvers less common, and make revolvers feel better to use.
Gunslinger should be funny and encourage cowboy blorbos to be more common.

image

image

## Changelog

:cl:
add: Gunslinger quirk! Lets you fling around a revolver like an old western! Good luck shooting anything newer, though.
add: 'New' HP Shadow sprites.
add: Double action revolvers have safeties now!
tweak: Many revolver tweaks
balance: Shadow and Ashhand are gate loaded now, and as such take longer to load.
fix: Non-speedloaders now load revolvers slowly again.
/:cl:

### CODE REVIEW

File: code/modules/projectiles/guns/gunhud.dm [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+
+/// Returns get_ammo() with the appropriate args passed to it - some guns like the revolver and bow are special cases
+/datum/component/ammo_hud/revolver/get_accurate_ammo_count(obj/item/gun/ballistic/revolver/the_gun)
+	if(istype(the_gun, /obj/item/gun/ballistic/revolver))
+		var/obj/item/gun/ballistic/revolver/the_revolver = the_gun
+		if(the_revolver.magazine)
+			return the_revolver.magazine.ammo_list()
+		else
+			return FALSE
```
Comment (west3436) [inline code/modules/projectiles/guns/gunhud.dm]: Probably can get away with killing this else branch and just allowing this proc to return null

File: code/modules/projectiles/guns/ballistic/revolver.dm [hunk 2]
```diff
@@ -20,59 +27,318 @@
...
+		REVOLVER_ROTATE_LEFT = image(icon = 'icons/mob/radial.dmi', icon_state = "radial_revolver_left"),
+		REVOLVER_ROTATE_RIGHT = image(icon = 'icons/mob/radial.dmi', icon_state = "radial_revolver_right"),
+		REVOLVER_AUTO_ROTATE_LEFT_LOADING = image(icon = 'icons/mob/radial.dmi', icon_state = "radial_revolver_auto_left"),
+		REVOLVER_EJECT_ALL = image(icon = 'icons/mob/radial.dmi', icon_state = "radial_revolver_eject_all"),
+		REVOLVER_EJECT_CURRENT = image(icon = 'icons/mob/radial.dmi', icon_state = "radial_revolver_eject_one"),
+		REVOLVER_AUTO_ROTATE_RIGHT_LOADING = image(icon = 'icons/mob/radial.dmi', icon_state = "radial_revolver_auto_right"),
+		)
+
+	var/image/editing_image
+	if(gate_load_direction == REVOLVER_AUTO_ROTATE_RIGHT_LOADING)
```
Comment (west3436) [inline code/modules/projectiles/guns/ballistic/revolver.dm]: Can simplify:
`editing_image = chamber_options[gate_load_direction]`
`editing_image.icon_state = "radial_revolver_auto_[gate_load_direction == REVOLVER_AUTO_ROTATE_RIGHT_LOADING ? "right":"left"]_on"`

**Comment by rye-rice** [comment]
Updated the PR desc with:

"To elaborate on gunslinger:

It gives you more recoil and spread when using SMGs, assault rifles, and LMGs, it increases accuracy with revolvers, shotguns, and lever actions, and no changes to anything else.

Additionally, when using some gate loaded revolvers, you will see a ammo hud representing the chambers and what or if a cartridge is loaded. You can alt click to open up a action menu, which will allow you to rotate the cylinder or empty the current gate round.

Also, lets you fan SINGLE ACTION revolvers (HP Shadow and HP Ashhand) by one handing it with nothing else in your offhand. Slightly faster than pistols to boot."

**Comment by PrefabQuasar** [comment]
a couple notes.
Double action revolvers feel Janky with this, for one, you can't load them from an ammo box At All, and have to take the bullets out of the box individually. It seems like double-actions are actually automatics playing pretend? pulling the trigger on an empty cylinder doesnt cycle the cylinders forward(it should, the trigger cocks the hammer when pulled, actuating the cylinder), and need the hammer to be "racked" by pressing space, which seems to act like an automatic slide(except for shell ejection, thankfully)
Additionally, single action revolvers seem to rotate the cylinder both on the hammer cocking *and* decocking, which is incorrect, they should only cycle forward on cocking

On a lesser note, I feel as though some of the abilities locked behind gunslinger should be available by default, being able to rotate the cylinder without using the hammer or trigger is Quite Basic Gun Use, and doesnt really make sense to be so restricted. The cylinder view should probably be visible to normal people on all revolvers BUT the gate-loaders, which could require the quirk. The only things that feel as though they genuinely Should come with the downsides gunslinger does are hammer fanning and the revolver tricks which are both actually advanced techniques.

**Comment by MarkSuckerberg** [comment]
ugh. quirks.

First of all, yeah, definitely don't lock actual features like that behind a quirk, I feel like the accuracy changes are already pushing it but I won't outright reject that unless I see it being abused. As much as it's funny and `quirky` I believe that should be visible to all players or none. Honestly I'd rather people just RP things and just use certain guns more instead of outright getting buffs to it, again, I'll let it slide.

Second of all, minor note and more to prefab here, isn't that supposed to be the case with ammo boxes? I assumed we'd removed the ability to autoload with ammo boxes like that so that reloading was less trivial, but I could be wrong on that

**Comment by rye-rice** [comment]
Updated gunslinger and made the ineractions available to everyone, and to answer mark and prefab, you have to unload the cylinder with your offhand first to load new bullets into it, instead of putting in a speedloader and it magically pushing out all the spent casings.

**Comment by rye-rice** [comment]
Conflicts fixed

**Comment by rye-rice** [comment]
conflic fixed

**Pull request merged**